lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/WordStorage.java

```diff
@@ -0,0 +1,338 @@
+          in.setPosition(pos);
+          for (int i = currentEntry.length() - 1; i >= commonPrefixLength; i--) {
+            char c = (char) in.readVInt();
+            assert c == currentEntry.charAt(i);
+            pos -= in.readVInt();
+            in.setPosition(pos);
+          }
+          commonPrefixPos = pos;
+        }
+        currentEntry = entry;
+      }
+
+      group.add(flags);
+      if (hasCustomMorphData) {
+        morphDataIDs.add(morphDataID);
+      }
+    }
+
+    private int flushGroup() throws IOException {
+      IntsRefBuilder currentOrds = new IntsRefBuilder();
+
+      boolean hasNonHidden = false;
+      for (char[] flags : group) {
+        if (!hasHiddenFlag(flags)) {
+          hasNonHidden = true;
+          break;
+        }
+      }
+
+      for (int i = 0; i < group.size(); i++) {
+        char[] flags = group.get(i);
+        if (hasNonHidden && hasHiddenFlag(flags)) {
+          continue;
+        }
+
+        currentOrds.append(flagEnumerator.add(flags));
+        if (hasCustomMorphData) {
+          currentOrds.append(morphDataIDs.get(i));
+        }
+      }
+
+      int lastPos = commonPrefixPos;
+      for (int i = commonPrefixLength; i < currentEntry.length() - 1; i++) {
+        int pos = dataWriter.getPosition();
+        ensureArraySize(0, false);
+        dataWriter.writeVInt(currentEntry.charAt(i));
+        dataWriter.writeVInt(pos - lastPos);
+        lastPos = pos;
+      }
+
+      int pos = dataWriter.getPosition();
+      int hash = Math.abs(currentEntry.hashCode() % hashTable.length);
+      int collision = hashTable[hash];
+      hashTable[hash] = collision == 0 ? pos : -pos;
+
+      if (++chainLengths[hash] > 20) {
+        throw new RuntimeException(
+            "Too many collisions, please report this to dev@lucene.apache.org");
+      }
+
+      ensureArraySize(currentOrds.length(), collision != 0);
+      dataWriter.writeVInt(currentEntry.charAt(currentEntry.length() - 1));
+      dataWriter.writeVInt(pos - lastPos);
+      IntSequenceOutputs.getSingleton().write(currentOrds.get(), dataWriter);
+      if (collision != 0) {
+        dataWriter.writeByte(collision < 0 ? (byte) 1 : 0);
+        dataWriter.writeVInt(pos - Math.abs(collision));
+      }
+
+      group.clear();
+      morphDataIDs.clear();
+      return pos;
+    }
+
+    private void ensureArraySize(int valueLength, boolean hasCollision) {
+      int pos = dataWriter.getPosition();
+      int maxEntrySize = 8 + 4 * (valueLength + 1) + (hasCollision ? 5 : 0);
+      while (wordData.length < pos + maxEntrySize) {
+        wordData = ArrayUtil.grow(wordData);
+        dataWriter.reset(wordData, pos, wordData.length - pos);
+      }
+    }
+
+    private static boolean hasHiddenFlag(char[] flags) {
+      for (char flag : flags) {
+        if (flag == Dictionary.HIDDEN_FLAG) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    WordStorage build() throws IOException {
+      flushGroup();
```

Comment:
Maybe confirm this is only called once, in the end?

Also, should we confirm that the up-front declared `wordCount` is exactly how many (unique) words we did see?

Reply:
> Maybe confirm this is only called once, in the end?

Will do, thanks!

> Also, should we confirm that the up-front declared wordCount is exactly how many (unique) words we did see?

Actually, it's the approximate non-unique word count. I'll document that.